src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs

```diff
@@ -0,0 +1,190 @@
+using System.Net;
+using Java.Nio;
+using Java.Nio.Channels;
+using Java.Util.Concurrent;
+using Microsoft.IO;
+using Xamarin.Chromium.CroNet;
+
+namespace ActualChat.App.Maui;
+
+public class CronetMessageHandler : HttpMessageHandler
+{
+    private readonly CronetEngine _cronetEngine;
+
+    private IExecutorService Executor { get; }
+
+    public CronetMessageHandler(IExecutorService executor)
+    {
+        Executor = executor;
+        _cronetEngine = new CronetEngine.Builder(MauiApplication.Current)
+            .EnableBrotli(true)
+            .EnableHttp2(true)
+            .EnableQuic(true)
+            .AddQuicHint("dev.actual.chat", 443, 443)
+            .AddQuicHint("media-dev.actual.chat", 443, 443)
+            .AddQuicHint("cdn-dev.actual.chat", 443, 443)
+            .AddQuicHint("actual.chat", 443, 443)
+            .AddQuicHint("media.actual.chat", 443, 443)
+            .AddQuicHint("cdn.actual.chat", 443, 443)
+            .Build();
+    }
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        Debug.Assert(request.RequestUri != null);
+
+        var callback = new CronetCallback(request, cancellationToken);
+        var requestBuilder = _cronetEngine.NewUrlRequestBuilder(
+                request.RequestUri.ToString(),
+                callback,
+                Executor)
+            .SetHttpMethod(request.Method.Method)
+            .SetPriority(UrlRequest.Builder.RequestPriorityMedium)
+            .DisableCache();
+
+        foreach (var header in request.Headers)
+            requestBuilder.AddHeader(header.Key, string.Join(',', header.Value));
+
+        if (request.Content != null)
+        {
+            var body = await request.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+            requestBuilder.SetUploadDataProvider(UploadDataProviders.Create(body), Executor);
+            foreach (var header in request.Content.Headers)
+                requestBuilder.AddHeader(header.Key, string.Join(',', header.Value));
+        }
+        requestBuilder.Build().Start();
+
+        return await callback.Task.WaitAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        base.Dispose(disposing);
+        _cronetEngine.Dispose();
+    }
+
+
+    private class CronetCallback : UrlRequest.Callback
+    {
+        private const int MaxRedirectCount = 3;
+        private static readonly RecyclableMemoryStreamManager MemoryStreamManager = new ();
+
+        private readonly TaskCompletionSource<HttpResponseMessage> _result;
```

Comment:
not sure about maui, but don't we TaskSource?

Reply:
Will try to switch...

Reply:
Yep :)